libs/wingc/src/visit.rs

```diff
@@ -0,0 +1,272 @@
+use crate::ast::{
+	ArgList, Class, Constructor, Expr, ExprKind, FunctionDefinition, InterpolatedStringPart, Literal, Reference, Scope,
+	Stmt, StmtKind,
+};
+
+/// Visitor pattern inspired by implementation from https://docs.rs/syn/latest/syn/visit/index.html
+///
+/// If you are implementing a visitor, you can delegate back to the
+/// default implementation to visit the children of a node by calling
+/// visit::visit_<node_type>(self, node);
+///
+/// For example:
+///
+/// ```ignore
+/// impl<'ast> Visit<'ast> for ExprVisitor {
+///   fn visit_item_fn(&mut self, exp: &'ast Expr) {
+///     println!("Expr with span={}", exp.span);
+///
+///     // Delegate to the default impl to visit any nested expressions.
+///     visit::visit_expr(self, exp);
+///   }
+/// }
+/// ```
+pub trait Visit<'ast> {
```

Comment:
I think this begs to be a rust macro that receives a list of AST node types and generates the entire trait. Not critical but if I'm write here lets at least add a TODO.

Reply:
Totally. It's definitely possible to code-generate this kind of code. This would also make it easier to maintain multiple traits, like separate a `VisitMut` trait variation that gives mutable access to each tree node instead of read-only access.